### PR TITLE
WFLY-9271 Create CommandDispatcher-based BroadcastEndpoint

### PR DESCRIPTION
**Pull request by pferraro**
Prerequisite for JGroups 4.0.x upgrade.

https://issues.jboss.org/browse/WFLY-9271
https://issues.jboss.org/browse/WFLY-9229

@jmesnil This is an extract of the requisite commits from https://github.com/wildfly/wildfly/pull/10665. Please review.

### CODE REVIEW

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java [hunk 1]
```diff
@@ -132,13 +133,8 @@ public Object handle(Message message) throws Exception {
                 if (context == null) return NoSuchService.INSTANCE;
```
Comment (jmesnil) [inline clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java]: I'm not sure we allow such statement...

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java [hunk 2]
```diff
@@ -132,13 +133,8 @@ public Object handle(Message message) throws Exception {
...
                 Command<Object, Object> command = (Command<Object, Object>) unmarshaller.readObject();
-                Callable<Optional<Object>> task = new Callable<Optional<Object>>() {
-                    @Override
-                    public Optional<Object> call() throws Exception {
-                        // Wrap in an Optional, since command execution might return null
-                        return Optional.ofNullable(command.execute(context.orElse(null)));
-                    }
-                };
+                // Wrap execution result in an Optional, since command execution might return null
+                ExceptionSupplier<Optional<Object>, Exception> task = () -> Optional.ofNullable(command.execute(context.orElse(null)));
```
Comment (jmesnil) [inline clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java]: I'm not sure `context.orElse(null)` brings anything as you tested context nullity just above

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/web/shared/SharedSessionFailoverTestCase.java [hunk 3]
```diff
@@ -57,6 +57,7 @@
  */
 @RunWith(Arquillian.class)
 @RunAsClient
+@org.junit.Ignore("WFLY-9428")
```
Comment (jmesnil) [inline testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/web/shared/SharedSessionFailoverTestCase.java]: the issue does not look related to messaging-activemq changes, right?

**Review by jmesnil — CHANGES_REQUESTED** [review]

**Comment by jmesnil** [comment]
The PR looks good.
However I flagged it as "hold" as the main requirement for the messaging-activemq subsystem is to upgrade Artemis 2.4.0 without any regression (through #10641).
Once Artemis is upgraded, we can merge this PR.

**Review by jmesnil — APPROVED** [review]

**Comment by pferraro** [comment]
@jmesnil  Thanks.

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java (hunk 1 above)
Comment (pferraro) [reply clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java]: On the contrary, this can easily be the case when using ForkChannels.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/web/shared/SharedSessionFailoverTestCase.java (hunk 3 above)
Comment (pferraro) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/web/shared/SharedSessionFailoverTestCase.java]: It is a consequence of https://issues.jboss.org/browse/WFLY-9229.
The root cause of WFLY-9428 is more complex - and is addressed by https://github.com/wildfly/wildfly/pull/10665/commits/b0da6886d0f4ac3e464d3a9624908829db3fe1f6

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java (hunk 2 above)
Comment (pferraro) [reply clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java]: You're confusing a null context with a non-null, but non-present context.  These are completely different things.
A null context means that there is no command dispatcher registered with the given ID.  An non-null, but empty context means that a command dispatcher is registered with the given ID, but it was created with a null context (which is completely valid).  e.g. CommandDispatcher<Void>.

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java (hunk 1 above)
Comment (jmesnil) [reply clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java]: sorry, what I mean is that I'm not sure we allow to have if statement without curly braces on a single line :)

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java (hunk 2 above)
Comment (jmesnil) [reply clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java]: Ok, I see the diffrence now.
Maybe this should be put in a comment in the code as it's easy to be quite confused about what "non-null, but non-present context" mean :)

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java (hunk 1 above)
Comment (pferraro) [reply clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java]: Oh - I see. :)
A quick search of the wildfly code base reveals 308 uses of 1 liners for "simple if" statements, and 183 in wildfly-core.

File: clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java (hunk 2 above)
Comment (pferraro) [reply clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java]: No problem - I'll add some comments to clarify.

Comment (pferraro) [reply clustering/server/src/main/java/org/wildfly/clustering/server/dispatcher/ChannelCommandDispatcherFactory.java]: Done.  See: https://github.com/wildfly/wildfly/pull/10678/commits/002ea7eb2b4833b793ec5aa5a6c2db71328fefbe

**Review by jmesnil — APPROVED** [review]

**Comment by jmesnil** [comment]
I'm removing the hold on this PR. We are still working on the Artemis integration and will require another version of it that fixes some regression observed in Artemis 2.4.0.

This present PR can be merged.
